Vinyl/Track.swift

```diff
@@ -14,14 +14,40 @@ typealias HTTPHeaders = [String : String]
 typealias Request = NSURLRequest
 
 struct Response {
-    let urlResponse: NSURLResponse
+    let urlResponse: NSHTTPURLResponse
     let body: NSData?
     let error: NSError?
+    
+    init(urlResponse: NSHTTPURLResponse, body: NSData? = nil, error: NSError? = nil) {
+        self.urlResponse = urlResponse
+        self.body = body
+        self.error = error
+    }
 }
 
 struct Track {
-    let request: Request
+    let request: Request?
     let response: Response
+    
+    init(request: Request, response: Response) {
+        self.request = request
+        self.response = response
+    }
+    
+    init(response: Response) {
+        
+        self.response = response
+        
+        guard
+            let urlString = response.urlResponse.URL?.absoluteString,
```

Comment:
I think this is impossible, we cannot have a response that does not contain a URL or have a malformed one. `NSHTTPURLResponse`'s initializer receives an instance of `NSURL` which is not an optional so we have some assurances about having an URL here.